index.d.ts

```diff
@@ -1,99 +1,487 @@
 
-  remove(): void;
+  get(callback?: (value: string, instance: AutoNumeric) => void): string;
 
+  getFormatted(
+    callback?: (value: string, instance: AutoNumeric) => void
+  ): string;
+
+  /**
+   * Returns a plain number value without formatting
+   */
   getNumber(
     callback: (value: number | null, instance: AutoNumeric) => void = null
   ): number | null;
 
   getNumericString(
-    callback: (value: string | null, instance: AutoNumeric) => void = null
+    callback?: (value: string | null, instance: AutoNumeric) => void = null
   ): string | null;
 
-  static getPredefinedOptions(): AutoNumericOptions;
+  getLocalized(callback: (value: string) => void): string;
+
+  reformat(): void;
+
+  unformat(): void;
+
+  unformatLocalized(forcedOutputFormat?: OutputFormatOption): void;
+
+  isPristine(): boolean;
+
+  select(): void;
+
+  selectNumber(): void;
+
+  selectInteger(): void;
+
+  selectDecimal(): void;
+
+  clear(reset?: boolean): void;
+
+  update(...options: Options[]): Input;
+
+  /**
+   * Remove the autoNumeric listeners from the element (previous name : 'destroy'). Keep the element content intact.
+   */
+  remove(): void;
+
+  /**
+   * Remove the autoNumeric listeners from the element, and reset its value to ''
+   */
+  wipe(): void;
+
+  /**
+   * Remove the autoNumeric listeners from the element, and delete the DOM element altogether
+   */
+  nuke(): void;
+
+  /**
+   * Return the DOM element reference of the autoNumeric-managed element
+   */
+  node(): HTMLInputElement;
+
+  parent(): HTMLElement;
+
+  detach(): void;
+
+  attach(otherAnElement: HTMLElement, reFormat?: boolean): void;
+
+  init(domeElement2: HTMLElement): Input;
+
+  form(forcedSearch?: boolean): HTMLFormElement;
+
+  formNumericString(): string;
+
+  formFormatted(): string;
+
+  formLocalized(forcedOutputFormat?: PredefinedLanguages): string;
+
+  formArrayNumericString(): HTMLInputElement[];
+
+  formArrayFormatted(): HTMLInputElement[];
+
+  formArrayLocalized(): HTMLInputElement[];
+
+  formJsonNumericString(): string;
+
+  formJsonFormatted(): string;
+
+  formJsonLocalized(): string;
+
+  formUnformat(): void;
+
+  formReformat(): void;
+
+  formSubmitArrayNumericString(callback: Function): Input;
+
+  formSubmitArrayFormatted(callback: Function): Input;
+
+  formSubmitArrayLocalized(callback: Function): Input;
+
+  formSubmitJsonNumericString(callback: Function): Input;
+
+  formSubmitJsonFormatted(callback: Function): Input;
+
+  formSubmitJsonLocalized(callback: Function): Input;
 }
 
-/* Expose types as well */
+/**
+ * Expose types as well
+ */
 declare namespace AutoNumeric {
-  export type OutputFormat = 'string' | 'number';
+  export type OutputFormatOption =
+    | "string"
+    | "number"
+    | "."
+    | "-."
+    | ","
+    | "-,"
+    | ".-"
+    | ",-"
+    | null;
+
+  export type CaretPositionOption =
+    | "start"
+    | "end"
+    | "decimalLeft"
+    | "decimalRight"
+    | "doNoForceCaretPosition";
```

Comment:
This should be `null` if I understood how Typescript type definitions works

Reply:
I took a look at the options. The type definitions aren't 100% accurate in that area, so this is a place we have to work on (later?). I think they are well enough so maybe no blocker for getting this merged?

---
Details, based on the `CaretPositionOption` discussed here, but can be applied to other options, too:

There are three areas where changes are needed:

1. Type for "real" caret positions

`doNoForceCaretPosition` is not a caret position, so it should to be removed from the`CaretPositionOption` definition:

```typescript
type CaretPosition =
    | "start"
    | "end"
    | "decimalLeft"
    | "decimalRight";
```
 Note that I additionally changed the name from `CaretPositionOption` to `CaretPosition` to make the increased separation to the options clearer.

2. Values that can be applied to `caretPositionOnFocus` in `Options`

`null` is an allowed value for `caretPositionOnFocus`, meaning "do not force caret position", so it should be added to the definition of the `caretPositionOnFocus` option:

```typescript
interface Options {
  // ...
  caretPositionOnFocus?: CaretPosition | null,
  // ...
}
```

Alternatively we could (re-)introduce a `CaretPositionOption` and use it here:

```typescript
type CaretPositionOption = CaretPosition | null;

interface Options {
  // ...
  caretPositionOnFocus?: CaretPositionOption;
  // ...
}
```

3. Predefined option aliases

There's an static member [`AutoNumeric.options`](https://github.com/autoNumeric/autoNumeric/blob/master/src/AutoNumericOptions.js) holding option aliases (or whole preset) which is **currently missing** in the TypeScript definition.

Possible solutions:

A: generic type definition

A generic object type could be used as a workaround until the real one is modeled: 

```typescript
static options: object;
```

We would have no type-safety here, so TypeScript won't complain when accessing nonexistent properties. But that's the same behavior as having no TS defintions at all, anyway.

B: Exactly modeled options:

```typescript
static options = {
  // ...
  caretPositionOnFocus: {
    start: 'start';
    end: 'end';
    decimalLeft: 'decimalLeft';
    decimalRight: 'decimalRight';
    doNoForceCaretPosition: null;
}
  // ...
}
```

(I think I'll would additionally have to add some `const` to each (or to the options object as a whole) to give TypeScript a hint that those strings a no strings but constants which are compatible e.g. to `CaretPosition`, but that's standard stuff and should be no problem)

The problem I have with this is that this is completely redundant with the JavaScript code base and that will be a constant source of problems.

Maybe it's possible to directly derive from/include the `AutoNumericOptions.js` in the TypeScript definition, but I'll would have to do further research.

C: Compromise: partially modeled options

As a trade-off between type safety and redundancy, we could express what keys exist and that the presets they're holding are compatible with each option they are representing, e.g.:

```typescript
static options = {
  // ...
  caretPositionOnFocus: Record<string, CaretPositionOption | null>;
  // ...
}
```

Reply:
Well, I'm not a typescript user so your choice will be better than mine.

However regarding code quality, I strive to make code that is 100% green in tests, with maximum coverage and more importantly, that is not in an "half-state".

This is why I asked insistently if there was a way to automate the generation of such definition since the AutoNumeric jsdoc is 100% accurate and up-to-date.
Your point `B` mentions that sync problem that would invariably happen over time (specially since I do not have any automated tests to check against that TS definition).

However, since it seems from the very first tries to add a TS definition to the project that's it's difficult and/or takes time to do, I'm willing to accept to merge that first version when corrected, given that the goal is still to make it perfectly synchronized in the near future with the current code base.

Reply:
Altough I have not much experience with complex JSDoc annotations, I suppose that auto-generated `.d.ts` files won't be as expressive as hand-crafted ones. **But** they could well be expressive enough for this project, I don't know.

Auto-generating is possible since the latest TypeScript 3.7 release, which can generate such `.d.ts` files from JSDoc-annotated JavaScript files, see the [release notes](https://devblogs.microsoft.com/typescript/announcing-typescript-3-7/#user-content---declaration-and--allowjs).

I could try this for the options and default options first, as those are the most common places for unnoticed deviations, and "link" that with the manually generated files. In a second step, I could try a fully automated generation and compare it against the manual crafted definitions to get a sense for the quality.

First tries were rather promising (I even found an error in your JSdoc due to this, just commited to this PR), but for `AutoNumericOptions.js` and `AutoNumericPredefinedOptions.js`, extending the autoNumeric class via `Object.defineProperty()` is essentially "hiding" their contents from the TypeScript compiler, resulting in definitions with just empty exports. So this needs some changes.

As I currently don't have the time to pursue this further, I would be very grateful if the PR could be merged in its current state. It's far better for TypeScript user than having no TS definition at all, as we aren't forced to use the whole module as a blackbox with type `any`.

---
Note regarding testing of the TypeScript definitions: there could be TypeScript-only tests, that are testing the type system only. See for example all tests at the DefinitelyTyped project, for example <https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/test/tsx.tsx>. The test is successful if TypeScript compiles the test without errors

This could also be applied to the existing tests (running them in unnecessary as long as they are run with JavaScript, since TypeScript throws no own errors at runtime)

Reply:
I unfortunately have close to no time to investigate the `--declaration` and `--allowJs` flags and typescript compilers, nor typescript tests.

It would be great if you could continue your work on that subject even after I merge your PR ;)